tests/js/client/shell/transaction/shell-transaction-cluster.js

```diff
@@ -367,9 +372,404 @@ function transactionReplicationOnFollowersSuite(dbParams) {
+
+      // Commit a transaction and expect everything to work
+      let obj = {
+        collections: {
+          write: [c.name(), distLike.name()]
+        }
+      };
+      let trx = null;
+      try {
+        trx = db._createTransaction(obj);
+        let tc1 = trx.collection(c.name());
+        tc1.save({_key: "foo"});
+        let tc2 = trx.collection(distLike.name());
+        tc2.save({_key: "bar"});
+      } catch (err) {
+        fail("Transaction failed with: " + JSON.stringify(err));
+      } finally {
+        if (trx) {
+          trx.abort();
+        }
+      }
+
+      let replication2Log = '';
+      if (isReplication2) {
+        let {logs} = dh.getCollectionShardsAndLogs(db, c);
+        let log = logs[0];
+        let entries = log.head(1000);
+        replication2Log = `Log entries: ${JSON.stringify(entries)}`;
+      }
+
+      // Check that no documents were inserted
+      for (let [shard, key] of shards) {
+        let localValues = {};
+        for (const [serverId, endpoint] of Object.entries(servers)) {
+          lh.waitFor(
+            dh.localKeyStatus(endpoint, dbn, shard, key, false));
+          localValues[serverId] = dh.getLocalValue(endpoint, dbn, shard, key);
+        }
+        for (const [serverId, res] of Object.entries(localValues)) {
+        assertTrue(res.code === 404,
+          `Expected 404 while reading key ${key} from ${serverId}/${dbn}/${shard}, ` +
+          `but the response was ${JSON.stringify(res)}. All responses: ${JSON.stringify(localValues)}` +
+          `\n${replication2Log}`);
+        }
+      }
+    },
+  };
+}
+
+/**
+ * This test suite checks what happens with abandoned transactions.
+ */
+function transactionReplication2AbandonmentSuite() {
+  'use strict';
+  const dbn = 'UnitTestsTransactionDatabase';
+  const cn = 'UnitTestsTransaction';
+  let cols = [];
+
+  const {setUpAll, tearDownAll, setUpAnd, tearDownAnd} =
+    lh.testHelperFunctions(dbn, {replicationVersion: "2"});
+
+  return {
+    setUpAll,
+    tearDownAll,
+    setUp: setUpAnd(() => {
+      cols = [];
+    }),
+    tearDown: tearDownAnd(() => {
+      for (let col of cols) {
+        col.drop();
+      }
+    }),
+
+    testAbandonedTransaction: function (testName) {
+      let col = db._create(`${cn}-${testName}`, {
+        numberOfShards: 1,
+        writeConcern: 2,
+        replicationFactor: 3
+      });
+      cols.push(col);
+
+      let {logs} = dh.getCollectionShardsAndLogs(db, col);
+      let log = logs[0];
+
+      // Start a new transaction
+      let obj = {
+        collections: {
+          write: [col.name()]
+        }
+      };
+      let trx;
+      try {
+        trx = db._createTransaction(obj);
+        let tc = trx.collection(col.name());
+        tc.save({_key: "foo"});
+      } catch (err) {
+        fail("Transaction failed with: " + JSON.stringify(err));
+      }
+
+      // Drop the collection while the transaction is still ongoing
+      col.drop();
+      cols = [];
```

Comment:
Aha:
This place is actually a good point for the above tearDownAll() suggestion.
Assume the drop() get's so far that the collection is essentially dropped, and then throws an error (e.g. connection interruption) then cols still has the collection and tearDownAll would again fail. Causing following tests to be in bad states.

Reply:
https://github.com/arangodb/arangodb/pull/20012/commits/8a727afbd44c826fa4bf4a1dc42ec2218957b10f